### PR TITLE
Fix #923 Add advanced colors

### LINKED ISSUE #923: Support other properties from the URL even if the UI does not like ntp_ ones
If I visit this URL

https://color.firefox.com/?theme=XQAAAAKEAwAAAAAAAABBqYhm849SCicxcUV5wYurLpubfikoJrYDaXef7eDwUAjshNvI2R1-qq64xJcRb6iFRYLtb_2uPCIluCEjcsDFhxGAKMrYS8NozTghHsnboe7DXFl_DJxNDhj3ISa2-Y1jhjzCKx_Owt6rCmfzvQKMUtPQkskDCwjepZb-j1ucSfa4-ID1aN-qQ9aiRpUq20b8egfX-XhyZAujr68cTFrCEODOce4x7AncvKoiLElwJNJ1pSIqMo1khLnCPZsqsa5CrfWMdOsQsor0B4Mw5i36fNepTl44I-ZelGkPO7gtegjdtF44Ezr96Tv4rktLrQS4Pjr1q2O0pNxCfUhGw4MaW9MmJUXoN8kbfZyDy1mFy3DMMzUU_dSiVEyZXw-qjJXxL-ukSoZpR-OX8Wrx4s3AI5_DugvZedKR5wPBr5k10dgUa-YplllcjaSZc8-Rf4KjirpFcAiraNM7-4OsuUslFyDh_uKTyg

I expect the UI to look like this

image

In reality it looks like this:
image

The problem seems to be here https://github.com/mozilla/FirefoxColor/blob/master/src/preset-themes/default.json

Would you accept a PR?
The reason I want this feature is that it would be nice for themer.dev to be able to generate more complete themes and share them trough color.firefox.com and not necesarily require an AMO account. Here is the relevant comment https://github.com/mjswensen/themer/issues/48#issuecomment-617338335

The list of props I'm currently interested in adding is:

bookmark_text
frame_inactive
icons_attention
icons
ntp_background
ntp_text
popup_highlight_text
popup_highlight
toolbar_field_highlight_text
popup_border
sidebar_border
sidebar_highlight_text
sidebar_highlight
sidebar_text
sidebar
tab_loading
tab_text
toolbar_bottom_separator
toolbar_field_border
toolbar_field_separator
toolbar_field_text_focus
toolbar_field_border_focus

They don't necessarily need to be in the web ui. I'll stick to whatever is now the default. All I want to do is the extension to be able to render an URL like the one above and create the full theme, not a partial one.

### PR DESCRIPTION
**Pull request by andreicristianpetcu**
color_reset

image

image

Screenshot from 2020-05-12 20-11-23

### CODE REVIEW

**Comment by andreicristianpetcu** [comment]
https://github.com/mozilla/FirefoxColor/issues/923

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 1]
```diff
@@ -88,14 +93,26 @@ class ThemeColorsEditor extends React.Component {
...
-                  <span className="theme-unit__label" title={colorLabels[name]}>
-                    {colorLabels[name]}
+                  {color ?
+                    <span
+                      className="theme-unit__swatch"
+                      style={{ backgroundColor: colorToCSS(color) }}
+                      title={labels[name]}
+                    />
+                    : <span className="theme-unit__default">
+                      <img src={iconDefault} />
```
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: image
done

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 2]
```diff
@@ -117,11 +134,19 @@ class ThemeColorsEditor extends React.Component {
           {!selectedColor && (
             <div className="theme-colors-editor__prompt">
               <div className="theme-colors-editor__prompt-arrow" />
-              <p>Pick a color to start customizing Firefox.</p>
+              <div className="theme-colors-editor__prompt-description" >
+                <p>Pick a color to start customizing Firefox.</p>
+                {advancedColors && (
+                  <div>
+                    <p>You need to have <a href="https://addons.mozilla.org/en-US/firefox/addon/firefox-color/">Firefox Color extension</a> installed since advanced colors don&#39t have a preveiw in the web UI.</p>
+                    <p>Learn more about these colors from the <a href="https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/theme#colors">theme API on MDN</a>.</p>
```
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: done

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 3]
```diff
@@ -117,11 +134,19 @@ class ThemeColorsEditor extends React.Component {
           {!selectedColor && (
             <div className="theme-colors-editor__prompt">
               <div className="theme-colors-editor__prompt-arrow" />
-              <p>Pick a color to start customizing Firefox.</p>
+              <div className="theme-colors-editor__prompt-description" >
+                <p>Pick a color to start customizing Firefox.</p>
+                {advancedColors && (
+                  <div>
+                    <p>You need to have <a href="https://addons.mozilla.org/en-US/firefox/addon/firefox-color/">Firefox Color extension</a> installed since advanced colors don&#39t have a preveiw in the web UI.</p>
```
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js:142]: image
image

Do I still keep the "Pick a color to start customizing Firefox" in the advanced tab? @Rob--W

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 4]
```diff
@@ -88,14 +93,26 @@ class ThemeColorsEditor extends React.Component {
...
+                    : <span className="theme-unit__default">
+                      <img src={iconDefault} />
+                    </span>}
+
+                  <span className="theme-unit__label" title={labels[name]}>
+                    {labels[name]}
                   </span>
+
+                  {advancedColors && color && (
+                    <span className="theme-unit__revert" onClick={ev => this.props.clearColor({ name })}>
```
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js:114]: Ahh.... I guess I need to look into this. https://github.com/casesandberg/react-color

[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: How about this?
color_reset

[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: I'm marking this as resolved

File: src/web/lib/components/ThemeColorsEditor/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: image

**Comment by andreicristianpetcu** [comment]
@Rob--W all the issues have been fixed :smile:

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 5]
```diff
@@ -54,16 +61,25 @@ class ThemeColorsEditor extends React.Component {
...
+
     // Select only the color properties from the theme.
-    const colorKeys = Object.keys(colors).filter(name => name in colorLabels);
+    const colorKeys = Object.keys(labels);
 
     // Dedupe colors for swatch presets
     const uniqueColorArray = [
       ...new Set(
-        colorKeys.map(name => {
+        Object.keys(Object.assign({}, colorLabels, advancedColorLabels)).map(name => {
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:82]: This is not wrong, but since the use of spreads are widespread in the code base, how about using it here too?

`{...colorLabels, ...advancedColorLabels}`

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 6]
```diff
@@ -88,14 +104,22 @@ class ThemeColorsEditor extends React.Component {
...
-                  <span className="theme-unit__label" title={colorLabels[name]}>
-                    {colorLabels[name]}
+                  {color ?
+                    <span
+                      className="theme-unit__swatch"
+                      style={{backgroundColor: colorToCSS(color)}}
+                      title={labels[name]}
+                    />
+                    : <span
+                      className="theme-unit__default"
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:114]: This element and its CSS are now basically copies of the swatch, except less accessibility-friendly (the focus ring is not visible when focused).

Revert this change, and modify the logic in the original swatch element to render the dashed appearance.

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 7]
```diff
@@ -104,24 +128,52 @@ class ThemeColorsEditor extends React.Component {
...
+                color={colors[selectedColor]}
+                width="270px"
+                disableAlpha={!colorsWithAlpha.includes(selectedColor)}
+                onChangeComplete={nextColor =>
+                  this.handleColorChange(selectedColor, nextColor)
+                }
+                presetColors={uniqueColorArray}
+              />
+              {advancedColors &&
+                <button className={"use-default"}
```
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: Does this look better? @Rob--W 
image

**Comment by andreicristianpetcu** [comment]
I fixed all the comments. Please take a look again when you have some time, @Rob--W

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 8]
```diff
@@ -90,12 +104,14 @@ class ThemeColorsEditor extends React.Component {
                 >
                   <span
                     className="theme-unit__swatch"
-                    style={{ backgroundColor: colorToCSS(color) }}
-                    title={colorLabels[name]}
+                    style={{backgroundColor: color ? colorToCSS(color) : "white", borderStyle: color ? "solid" : "dashed"}}
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:107]: `"solid"` -> `""` to fall back to the default stylesheet.

But why not this instead?

```
color ? {backgroundColor: ...} : {borderStyle: "dashed"}
```

(code golfing isn't necessary; if readability improves by using multiple lines or even a separate variable, by all means use it.)

File: src/web/lib/components/ThemeColorsEditor/index.scss [hunk 9]
```diff
@@ -96,3 +103,25 @@
...
+  border-radius: 2px;
+  border: 1px dashed var(--grey-90-a30);
+  height: 16px;
+  width: 16px;
+  margin-left: 2px;
+  margin-right: 10px;
+  background-color: white;
+}
+
+button::-moz-focus-inner {
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.scss:125]: This is an accessibility feature. Don't change the behavior unless absolutely necessary. Why would you remove the border?

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 10]
```diff
@@ -104,20 +120,51 @@ class ThemeColorsEditor extends React.Component {
...
+                color={colors[selectedColor]}
+                width="270px"
+                disableAlpha={!colorsWithAlpha.includes(selectedColor)}
+                onChangeComplete={nextColor =>
+                  this.handleColorChange(selectedColor, nextColor)
+                }
+                presetColors={uniqueColorArray}
+              />
+              {advancedColors &&
+              <div className="use-default"
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:134]: Why is this a div with an embedded span? It almost seems like you tried to implement the appearance from the ASCII sketch at https://github.com/mozilla/FirefoxColor/pull/929#discussion_r424025891 . Is that the case? If so, let me clarify: The `[ ]` in the sketch should not be interpreted as a box (i.e. what you've put in the `<span/>` below). Rather, it's used to illustrate the behavior of standard UI elements (checkbox).

The curent HTML/CSS has the following problems:
- Due to the use of `div` (without tabindex), the element is not focusable and lacks any visual cues that the element would do anything (`:hover`).
- There is no obvious visual indication that the default color is used, or that the color picker is in use. One potential way to resolve that is to make the "Use default color" a label of a checkbox, and toggle the visibility of the color picker upon (un)checking it. There may be other ways.

File: src/web/lib/components/ThemeColorsEditor/backspace.svg [hunk 11]
```diff
@@ -0,0 +1,4 @@
+<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="backspace" class="svg-inline--fa fa-backspace fa-w-20" role="img"
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/backspace.svg:1]: This file is unused.

File: src/web/lib/components/ThemeColorsEditor/index.scss [hunk 12]
```diff
@@ -96,3 +103,25 @@
...
+
+.use-default {
+  display: inline-flex;
+  background-color: white;
+  border: none;
+  cursor: pointer;
+  margin-left: 8px;
+}
+
+.use-default span {
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.scss:115]: The `span` appears to only exist for styling purposes. If that's the case, consider using the `::before` pseudo-element instead.

File: src/web/lib/components/ThemeColorsEditor/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: hard to make the color picker disabled. is it ok if I use radio buttons instead of checkboxes? @Rob--W

File: src/web/lib/components/ThemeColorsEditor/index.js (hunk 10 above)
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: I have a radio proposal
radio
And a checkbox proposal.
checkbox
In both I have the same issue: what do I do when the user clicks "other" or unselects the "use default"? Currently it does nothing.

For the checkbox proposal I have the following issue: I don't know how to disable the color picker, the "visual cue that it is not in use".

Do you prefer one of these? I have separated commits that you can check out and try.

https://github.com/andreicristianpetcu/FirefoxColor/commit/0c8b59cf660c0f87a2aa1cd2317826e362f81027 and https://github.com/andreicristianpetcu/FirefoxColor/commit/a34db2755f442f01b14e7e612b0b016f18ca42ce

**Comment by andreicristianpetcu** [comment]
@Rob--W I fixed undo and redo. https://github.com/mozilla/FirefoxColor/pull/929/commits/bcde4494dcdeae520cec7f0ec2b5ebdc4bd117f6

**Comment by andreicristianpetcu** [comment]
@Rob--W can you please tell me when you would have some time to look into my PR? You usually respond fast but now there have been several days without reply. Thank you!

**Comment by Rob--W** [comment]
Quoted from https://github.com/mozilla/FirefoxColor/pull/929#discussion_r424714658
> I have a radio proposal
> [radio image link](https://user-images.githubusercontent.com/922368/81861958-aaf13900-9571-11ea-8d3d-10941814989b.gif)
> And a checkbox proposal.
> [checkbox image link](https://user-images.githubusercontent.com/922368/81861985-b3497400-9571-11ea-8b47-bb32c829d8d0.gif)
> In both I have the same issue: what do I do when the user clicks "other" or unselects the "use default"? Currently it does nothing.

When the color picker is always visible, the radio looks like a better choice, as it allows the user to re-activate the color in the color picker by clicking on it.

When the color picker is disabled / hidden when not in use, the checkbox is better, because there is no point in having to click twice to pick a color (first click on radio, then click on color picker)

> For the checkbox proposal I have the following issue: I don't know how to disable the color picker, the "visual cue that it is not in use".

Even without any support from the color picker library itself, there are multiple options available, in roughly two categories:

- Hiding it - HTML `[hidden]` attribute, CSS display none, ...
- Making it visually disabled through CSS:  opacity or maybe even [filter](https://developer.mozilla.org/en-US/docs/Web/CSS/filter) to finish the visual. If you really want to, you could use `pointer-events:none` to disable interaction.

In this case, I suppose that the most obvious way to offer the choice is as follows:

- Use the radio inputs (but please expand the radio label - I think that I originally suggested a longer, unambiguous text label)
- When color picker is deactivated: Render as grayscale with some opacity.
- When color picker is dectivated, but user hovers over it: adjust the appearance so that the user can see that they can interact with it, but still make it obvious that the chosen value is not used yet (e.g. by a bit of grayscale or a bit of transparency)

**Comment by andreicristianpetcu** [comment]
ff_color
@Rob--W I updated the text from the longer version of https://github.com/mozilla/FirefoxColor/pull/929#discussion_r424025891

image

**Comment by andreicristianpetcu** [comment]
If you have any other feedback, please be specific. "I want X" not "you can do either X or Y". It helps me focus on what I need to do. When I try to interpret the options I have I seem to postpone the merging of this PR.

**Comment by harogaston** [comment]
Just wanted to say great work man. I've been waiting for this since Firefox Color!

**Comment by Rob--W** [comment]
Sorry for the late review, I'm caught in other high priority work. I'll try to review by next week.

**Comment by andreicristianpetcu** [comment]
No problem. Thank you for doing this. @Rob--W

**Comment by andreicristianpetcu** [comment]
kind reminder @Rob--W

File: src/lib/constants.js [hunk 13]
```diff
@@ -33,6 +33,38 @@ export const colorLabels = {
   popup_text: "Popup Text"
 };
 
+export const advancedColorLabels = {
```
Comment (Rob--W) [inline src/lib/constants.js:36]: Just for the record: I verified that this list is complete. Every color listed here exists, and together with the existing `colorLabels` definition, every non-deprecated color property that Firefox [supports](https://searchfox.org/mozilla-central/rev/8ccea36c4fb09412609fb738c722830d7098602b/toolkit/components/extensions/schemas/theme.json#101-271) is covered.

File: src/lib/store-test.js [hunk 14]
```diff
@@ -233,4 +233,19 @@ describe("lib/store", () => {
       expect(result()).to.deep.equal(["4", "0", "2", "3", "1"]);
     });
   });
+
+  describe("colors backgrounds", () => {
+    it("should set advance colors", () => {
```
Comment (Rob--W) [inline src/lib/store-test.js:238]: advance -> advanced.

But this would result in redundancy after following my previous suggestion of renaming the `description` label. Let's change it to `should set ntp_background`

File: src/lib/store-test.js [hunk 15]
```diff
@@ -233,4 +233,19 @@ describe("lib/store", () => {
...
+      store.dispatch(actions.theme.setColor({ name: "ntp_background", color: {r: 255, g: 0, b: 0} }));
+
+      expect(store.getState().theme.present.colors.ntp_background).to.deep.equal({r: 255, g: 0, b: 0});
+    });
+
+    it("should set clear colors", () => {
+      store.dispatch(actions.theme.setColor({ name: "ntp_background", color: {r: 255, g: 0, b: 0} }));
+      store.dispatch(actions.theme.clearColor({ name: "ntp_background"}));
+
+      expect(store.getState().theme.present.colors.ntp_background).to.be.an("undefined");
```
Comment (Rob--W) [inline src/lib/store-test.js:248]: This check should be changed, to verify that the key doesn't exist at all.

File: src/lib/store-test.js [hunk 16]
```diff
@@ -233,4 +233,19 @@ describe("lib/store", () => {
       expect(result()).to.deep.equal(["4", "0", "2", "3", "1"]);
     });
   });
+
+  describe("colors backgrounds", () => {
```
Comment (Rob--W) [inline src/lib/store-test.js:237]: "colors background" -> Managing advanced colors

File: src/lib/store-test.js [hunk 17]
```diff
@@ -233,4 +233,19 @@ describe("lib/store", () => {
...
+    it("should set advance colors", () => {
+      store.dispatch(actions.theme.setColor({ name: "ntp_background", color: {r: 255, g: 0, b: 0} }));
+
+      expect(store.getState().theme.present.colors.ntp_background).to.deep.equal({r: 255, g: 0, b: 0});
+    });
+
+    it("should set clear colors", () => {
```
Comment (Rob--W) [inline src/lib/store-test.js:244]: The text should not be a bunch of keywords, but ideally form a coherent sentence.

How about: `it("should set clear colors"` -> `it("should clear ntp_background"` ?

File: src/lib/themes-test.js [hunk 18]
```diff
@@ -93,6 +93,30 @@ describe("lib/themes", () => {
...
+        "toolbar_vertical_separator"
+      ];
+      advancedColors.forEach(advancedColor => {
+        const theme = {
+          colors: {
+            ...presetThemesContext("./default.json").colors,
+          },
+          title: ["custom theme"],
+        };
+        theme.colors[advancedColor] = { r: 12, g: 34, b: 56 };
```
Comment (Rob--W) [inline src/lib/themes-test.js:113]: Use [computed property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#Computed_property_names) instead of `theme.colors[advancedColor] = ...` :
```
        const theme = {
          colors: {
            ...presetThemesContext("./default.json").colors,
            [advancedColor]: { r: 12, g: 34, b: 56 },
          },
          title: ["custom theme"],
        };
```

File: src/lib/themes-test.js [hunk 19]
```diff
@@ -93,6 +93,30 @@ describe("lib/themes", () => {
       // Input object should not have been changed.
       expect(themeInput).to.deep.equal(deprecatedTheme);
     });
+
+    it("should include advanced colors themes", () => {
```
Comment (Rob--W) [inline src/lib/themes-test.js:97]: To verify that the test below is not just passing due to blindly accepting any color property, could you add a test that includes an unsupported property name, and verify the resulting behavior?

File: src/lib/themes-test.js [hunk 20]
```diff
@@ -223,6 +247,15 @@ describe("lib/themes", () => {
       ).to.be.false;
     });
 
+    it("should reject difference in number of colors", () => {
+      expect(
+        subject(
```
Comment (Rob--W) [inline src/lib/themes-test.js:252]: The original comparison code for which you've written this test is not failing due to the incorrect number of colors, but due to not recognizing unsupported colors.

To make the test complete:

- Also add a test where the compared operands have the same properties, but differing only in the value of the advanced color. 
- Add a test where an unsupported color property is compared. I think that the expected result is that unrecognized colors are ignored.

File: src/lib/themes.js [hunk 21]
```diff
@@ -61,6 +62,10 @@ export const themesEqual = (themeA, themeB) => {
     return true;
   }
 
+  if (Object.keys(themeA.colors).length !== Object.keys(themeB.colors).length) {
```
Comment (Rob--W) [inline src/lib/themes.js:65]: As mentioned in the test, this is not the right check. You should be checking default colors + advanced colors, if any.

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 22]
```diff
@@ -43,6 +48,15 @@ class ThemeColorsEditor extends React.Component {
     this.props.setColor({ name, color: color.rgb });
   };
 
+  selectLastColor = (name) => {
+    const { theme: {colors}} = this.props;
+    this.props.setColor({name, color: this.colorPicker.state.rgb || colors.toolbar});
```
[the remark this answers is not in the document]
Comment (andreicristianpetcu) [reply src/web/lib/components/ThemeColorsEditor/index.js]: I wanted to use the last selected color from the picker but it does not quite work. I fixed it by having a field that tracks all the last colors.

**Comment by andreicristianpetcu** [comment]
@Rob--W DONE :smile:

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 23]
```diff
@@ -13,6 +14,14 @@ const DISMISS_CLASSNAMES = ["color__label", "color__swatch"];
 class ThemeColorsEditor extends React.Component {
   constructor(props) {
     super(props);
+    this.colorPicker = null;
+    this.setColorPickerRef = element => {
+      this.colorPicker = element;
+    };
+    const colors = this.props.theme.colors;
+    this.lastSelectedColors = {
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:22]: There is no need for storing the last known colors on all individual themes. The value would be lost anyway when the page refreshes.

There is only a need to remember one color, because otherwise the color picker became jumpy. See my comment below.

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 24]
```diff
@@ -90,34 +117,75 @@ class ThemeColorsEditor extends React.Component {
...
+                          this.selectLastColor(selectedColor)
+                        }/>
+                or select a color:
+              </label>
+            </div>           
+          }
+          {selectedColor &&
             <SketchPicker
+              className={advancedColors && !colors[selectedColor] ? "theme-colors-editor__disabled" : ""}
               color={colors[selectedColor]}
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:154]: The color picker is still jumping. Use this to fix that:

`colors[selectedColor] || this.lastSelectedColor`

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 25]
```diff
@@ -90,34 +117,75 @@ class ThemeColorsEditor extends React.Component {
...
+            <div className="theme-colors-editor__options">
+              <label>
+                <input type="radio" value="default" checked={!colors[selectedColor]}
+                        onChange={ev => this.handleClearColor(selectedColor)}/>
+                use Firefox&#39;s default style for this color
+              </label>
+              <label>
+                <input type="radio" value="other" checked={!!colors[selectedColor]}
+                        onChange={ev =>
+                          this.selectLastColor(selectedColor)
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:145]: `                          this.handleColorChange(this.lastSelectedColor)`

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 26]
```diff
@@ -13,6 +14,14 @@ const DISMISS_CLASSNAMES = ["color__label", "color__swatch"];
 class ThemeColorsEditor extends React.Component {
   constructor(props) {
     super(props);
+    this.colorPicker = null;
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:17]: colorPicker / colorPickerRef is not used.

**Comment by andreicristianpetcu** [comment]
@Rob--W done :rainbow:

File: src/web/lib/components/ThemeColorsEditor/index.js [hunk 27]
```diff
@@ -40,7 +41,14 @@ class ThemeColorsEditor extends React.Component {
...
 
   handleColorChange = (name, color) => {
-    this.props.setColor({ name, color: color.rgb });
+    this.lastSelectedColor = color;
+    this.props.setColor({ name, color });
+  };
+
+  handleClearColor = (name) => {
+    const { theme: {colors} } = this.props;
+    this.lastSelectedColor = colors[name];
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.js:50]: Instead of assigning `lastSelectedColor` on change, assign it on render. This has several benefits:
- less code duplication.
- also works if props are changed by an external component.
- when switching between a color with a picker to a "default" color, the color picker will retain its color (instead of the color of the last click).

Something like this:

```diff
diff --git a/src/web/lib/components/ThemeColorsEditor/index.js b/src/web/lib/components/ThemeColorsEditor/index.js
index 2963e2f..6a19588 100644
--- a/src/web/lib/components/ThemeColorsEditor/index.js
+++ b/src/web/lib/components/ThemeColorsEditor/index.js
@@ -41,13 +41,10 @@ class ThemeColorsEditor extends React.Component {
   };
 
   handleColorChange = (name, color) => {
-    this.lastSelectedColor = color;
     this.props.setColor({ name, color });
   };
 
   handleClearColor = (name) => {
-    const { theme: {colors} } = this.props;
-    this.lastSelectedColor = colors[name];
     this.props.clearColor({name});
   };
 
@@ -86,6 +83,11 @@ class ThemeColorsEditor extends React.Component {
       )
     ];
 
+    const selectedColorValue = colors[selectedColor];
+    if (selectedColorValue) {
+      this.lastSelectedColor = selectedColorValue;
+    }
+
     return (
       <div className="theme-colors-editor">
         <div className="theme-colors-editor-main">
@@ -125,12 +127,12 @@ class ThemeColorsEditor extends React.Component {
           {selectedColor && advancedColors && 
             <div className="theme-colors-editor__options">
               <label>
-                <input type="radio" value="default" checked={!colors[selectedColor]}
+                <input type="radio" value="default" checked={!selectedColorValue}
                         onChange={ev => this.handleClearColor(selectedColor)}/>
                 use Firefox&#39;s default style for this color
               </label>
               <label>
-                <input type="radio" value="other" checked={!!colors[selectedColor]}
+                <input type="radio" value="other" checked={!!selectedColorValue}
                         onChange={ev =>
                           this.handleColorChange(selectedColor, this.lastSelectedColor)
                         }/>
@@ -140,8 +142,8 @@ class ThemeColorsEditor extends React.Component {
           }
           {selectedColor &&
             <SketchPicker
-              className={advancedColors && !colors[selectedColor] ? "theme-colors-editor__disabled" : ""}
-              color={colors[selectedColor] || this.lastSelectedColor}
+              className={advancedColors && !selectedColorValue ? "theme-colors-editor__disabled" : ""}
+              color={selectedColorValue || this.lastSelectedColor}
               width="270px"
               disableAlpha={!colorsWithAlpha.includes(selectedColor)}
               onChangeComplete={nextColor =>

File: src/web/lib/components/ThemeColorsEditor/index.scss [hunk 28]
```diff
@@ -49,23 +49,31 @@
...
-
   .theme-unit__swatch {
     border-radius: 2px;
     border: 1px solid var(--grey-90-a30);
     height: 20px;
     width: 20px;
     margin-right: $grid * 4;
   }
+
+  .theme-unit__revert {
```
Comment (Rob--W) [inline src/web/lib/components/ThemeColorsEditor/index.scss:60]: Unused, remove.

**Comment by andreicristianpetcu** [comment]
@Rob--W done :tada:

**Comment by Rob--W** [comment]
Thanks for your work here. I've approved it, and will merge it once it's clear how/when we are updating the add-on. We don't want to publish the website with this feature without actually having the add-on present.

Because of an upcoming event, I'm not expecting a release in the next week.

**Comment by andreicristianpetcu** [comment]
@Rob--W when will you have some time to release this new version?

**Comment by Rob--W** [comment]
I have some time this week. To make sure that everything works (not just your patch, but also other changes since the push to production), I've sent a request to QA for testing.

**Comment by Rob--W** [comment]
QA will test in two weeks, so I'm expecting to actually release an update in the week of 6 July.

**Comment by andreicristianpetcu** [comment]
Is there a bugzilla I can track? @Rob--W

**Comment by Rob--W** [comment]
There is no separate Bugzilla ticket. QA finished verification and found some issues, reported at https://github.com/mozilla/FirefoxColor/issues?q=is%3Aissue+is%3Aopen+933+934+935+936+937+938+939

Some are unrelated to your feature (the test scope included the new "Revert" feature), but others are. They need to be investigated (to see if it's issue with Firefox Color, or maybe even Firefox itself):

- #936 - broken feature?
- #934 - small polish
- #933 - broken feature?

Can you take a look at the above? I don't have time to look at those, the most I can do this week is to publish a new version.
